✨ zb: add ProxyImpl trait

DerZade opened the pull request:
This PR fixes #433 by adding a `ProxyImpl` trait for macro generated proxy-structs to allow narrowing generic arguments to D-Bus proxies.

zeenix commented on zbus/src/blocking/proxy/mod.rs:
```diff
@@ -527,3 +527,19 @@ mod tests {
 }
+
+/// This trait is implemented by all blocking proxies, which were
+/// generated with the [`dbus_proxy`](zbus::dbus_proxy) macro.
```
Does this way of linking works? :thinking:

DerZade replied:
![chrome_sl6iP9qlT1](https://github.com/dbus2/zbus/assets/12705416/6591eeac-87f2-43e3-9f72-1abc3bfb93ef)


Yes it does 🙃

zeenix replied:
Nice. TIL :)

zeenix commented on zbus/src/lib.rs:
```diff
@@ -231,10 +231,11 @@ mod tests {
 
     use crate::{
-        blocking::{self, MessageIterator},
+        blocking::{self, proxy::ProxyImpl as _, MessageIterator},
```
don't think we need these changes here.

DerZade replied:
Considering your other comment, you are correct.

zeenix reviewed (changes requested):
Looking good otherwise. Also two nits:

* it'd be best to split the commit into two: One to add the new trait and another to implement it in the macros.
* I really appreciate you providing a nice commit log with ref to the issue. :+1:  Just one nit: Please break the lines at 74 chars as per the [guidelines](https://wiki.gnome.org/Git/CommitMessages).

DerZade commented on zbus/src/fdo.rs:
```diff
@@ -17,8 +17,8 @@ use zvariant::{
 use crate::{
-    dbus_interface, dbus_proxy, message::Header, object_server::SignalContext, DBusError, Guid,
-    ObjectServer,
+    dbus_interface, dbus_proxy, message::Header, object_server::SignalContext, proxy::ProxyImpl,
+    DBusError, Guid, ObjectServer,
```
That's actually really smart. I've never seen that pattern.

I actually thought about how I could circumvent existing code from breaking because they have to import the trait, but couldn't come up with a nice solution.

DerZade commented:
Thanks a lot for the amazingly fast review.

Sorry, apparently I missed the note about the 75 character limit. That is now corrected along with all your other comments. I also split the commits into two as requested 😇 

Let me know if there is anything else that needs to be addressed.

zeenix reviewed (approved):
LGTM. Thanks so much.
